nomad/state/state_store.go

```diff
@@ -4125,6 +4126,45 @@ func (s *StateStore) upsertAllocsImpl(index uint64, allocs []*structs.Allocation
 			if alloc.Job == nil {
 				return fmt.Errorf("attempting to upsert allocation %q without a job", alloc.ID)
 			}
+
+			// Check if the alloc requires sticky volumes. If yes, find a node
+			// that has the right volume and update the task group volume
+			// claims table
+			stickyVolumes := []*structs.TaskGroupVolumeClaim{}
+			for _, tg := range alloc.Job.TaskGroups {
+				for _, v := range tg.Volumes {
+					if !v.Sticky {
+						continue
+					}
+					stickyVolumes = append(stickyVolumes, &structs.TaskGroupVolumeClaim{
+						ID:            uuid.Generate(),
+						JobID:         alloc.JobID,
+						TaskGroupName: tg.Name,
+						AllocID:       alloc.ID,
+						VolumeName:    v.Source,
+					})
+				}
+			}
+
+			if len(stickyVolumes) > 0 {
+				allocNode, err := s.NodeByID(nil, alloc.NodeID)
+				if err != nil {
+					return err
+				}
+
+				for _, v := range allocNode.HostVolumes {
+					// Record volumes that this allocation uses in the claims table
+					for _, sv := range stickyVolumes {
+						if sv.VolumeName != v.Name {
+							continue
+						}
+						sv.VolumeID = v.ID
+						if err := s.UpsertTaskGroupVolumeClaim(index, sv); err != nil {
```

Comment:
Note that we've already got a transaction created (it's passed as an argument to `upsertAllocsImpl`, so I think we want this method to take that transaction rather than creating its own. That way we commit all the changes atomically.